module/builder/consensus/builder.go

```diff
@@ -133,6 +139,111 @@ func (b *Builder) BuildOn(parentID flow.Identifier, setter func(*flow.Header) er
 	return proposal.Header, nil
 }
 
+// repopulateExecutionTree restores latest state of execution tree mempool based on local chain state information.
+// Repopulating of execution tree is split into two parts:
+// 1) traverse backwards all finalized blocks starting from last finalized block till we reach last sealed block. [lastSealedHeight, lastFinalizedHeight]
+// 2) traverse forward all unfinalized(pending) blocks starting from last finalized block.
+// For each block that is being traversed we will collect execution results and add them to execution tree.
+func (b *Builder) repopulateExecutionTree() error {
+	finalizedSnapshot := b.state.Final()
+	finalized, err := finalizedSnapshot.Head()
+	if err != nil {
+		return fmt.Errorf("could not retrieve finalized block: %w", err)
+	}
+	finalizedID := finalized.ID()
+
+	// Get the latest sealed block on this fork, ie the highest block for which
+	// there is a seal in this fork.
+	latestSeal, err := b.seals.ByBlockID(finalizedID)
+	if err != nil {
+		return fmt.Errorf("could not retrieve parent seal (%x): %w", finalizedID, err)
+	}
+
+	latestSealed, err := b.headers.ByBlockID(latestSeal.BlockID)
+	if err != nil {
+		return fmt.Errorf("could not retrieve latest sealed block (%x): %w", latestSeal.BlockID, err)
+	}
+
+	sealedResult, err := b.resultsDB.ByID(latestSeal.ResultID)
+	if err != nil {
+		return fmt.Errorf("could not retrieve sealed result (%x): %w", latestSeal.ResultID, err)
+	}
+
+	// handle latest sealed result manually
+	// this is needed to handle a case when we are building on top of genesis block
+	err = b.recPool.AddResult(sealedResult, latestSealed)
+	if err != nil {
+		return fmt.Errorf("failed to add sealed result as vertex to ExecutionTree (%x): %w", latestSeal.ResultID, err)
+	}
+
+	blockLookup := make(map[flow.Identifier]*flow.Header)
+
+	// usually we start with empty execution tree, prune it to minimum height before adding results
+	err = b.recPool.PruneUpToHeight(latestSealed.Height)
+	if err != nil {
+		return fmt.Errorf("could not prune execution tree to height %d: %w", latestSealed.Height, err)
+	}
+
+	// traverse block and collect result IDs that were included in that block
+	traverser := func(header *flow.Header) error {
+		index, err := b.index.ByBlockID(header.ID())
+		if err != nil {
+			return fmt.Errorf("could not retrieve index for block (%x): %w", header.ID(), err)
+		}
+
+		for _, resultID := range index.ResultIDs {
```

Comment:
⚠️ 
The execution tree is supposed to hold _all_ known receipts for unsealed blocks. Actually the receipts _not_ incorporated into any blocks are the most relevant, as those should be included in future blocks. 


I would suggest to add _all known_ receipts for the respective block. We currently do this in `sealing.Core` https://github.com/onflow/flow-go/blob/5433e1d3959767dc5e1f6b3ae28092502e8308b9/engine/consensus/sealing/core.go#L975-L1000 but I suspect this is subject to change (as indicated in the comment). 

#### Comment 1

I have implemented the bulk of this logic in my PR. 

#### Comment 2

I think this part of the comment in `sealing.Core` is outdated: https://github.com/onflow/flow-go/blob/5433e1d3959767dc5e1f6b3ae28092502e8308b9/engine/consensus/sealing/core.go#L982-L985 receipts are added as soon as we store a block.